src/checkers/inference/model/serialization/ToStringSerializer.java

```diff
@@ -24,49 +28,65 @@
 import checkers.inference.model.VariableSlot;
 
 /**
- * This Serializer is meant only to convert constraints and variables to
- * human readable strings.  It is used currently by the DebugSolver.
+ * This Serializer is meant only to convert constraints and variables to human readable strings. It
+ * is used currently by the DebugSolver.
  */
 public class ToStringSerializer implements Serializer<String, String> {
     private final boolean showAstPaths;
     private boolean showVerboseVars;
 
-    private int indent = 0;
-    public static final String INDENT_STRING = "    ";
+    private int indentationLevel = 0;
+
+    // 2 spaces per indentation
+    public static final String INDENT = "  ";
+
+    // stores N concatenations of INDENT at index N, where index 0 stores empty string
+    // the index is the indentation level
+    protected final List<String> indentStrings = new ArrayList<>();
+
+    // used to format constant slots
+    protected final AnnotationFormatter formatter;
 
     public ToStringSerializer(boolean showAstPaths) {
         this.showAstPaths = showAstPaths;
         this.showVerboseVars = true;
+        formatter = InferenceMain.getInstance().getRealTypeFactory().getAnnotationFormatter();
+
+        // set first value to ""
+        indentStrings.add("");
     }
 
-    public void setIndent(int indent) {
-        this.indent = indent;
+    public void setIndentationLevel(int indentationLevel) {
```

Comment:
Instead of letting any `indentationLevel` is passed, can we just provide method, for example, `increate/decreaseIndentationLevel()` and hide the updating logic of `indentationLevel` so that we have more control at the owner side instead of trusting clients to pass a valid `indentationLevel`?
Can't we update `indentationStrings` list to add missing strings for the possible maximum `indentationLevel` in this method(or the new `increateIndentationLevel()`)? For me, it's natural to update the list here instead of in `indent()` method.

Reply:
There's one call sit to this method in DebugSolver, setting indentation level to 1.

Will move `indentationStrings` update logic to here. Thanks for the catch!